# Финальный проект. SQL. Анализ базы данных книжного сервиса

**Описание проекта**

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Задачи проекта:**

* посчитать, сколько книг вышло после 1 января 2000 года;
* для каждой книги посчитать количество обзоров и среднюю оценку;
* определить издательство, которое выпустило наибольшее число книг толще 50 страниц — (исключение из анализа брошюр);
* определить автора с самой высокой средней оценкой книг, учитывая только книги с 50 и более оценками;
* посчитайть среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

**Описание данных**

**Таблица `books`**

Содержит данные о книгах:

* `book_id` — идентификатор книги;
* `author_id` — идентификатор автора;
* `title` — название книги;
* `num_pages` — количество страниц;
* `publication_date` — дата публикации книги;
* `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

* `author_id` — идентификатор автора;
* `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

* `publisher_id` — идентификатор издательства;
* `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

* `rating_id` — идентификатор оценки;
* `book_id` — идентификатор книги;
* `username` — имя пользователя, оставившего оценку;
* `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах:

* `review_id` — идентификатор обзора;
* `book_id` — идентификатор книги;
* `username` — имя автора обзора;
* `text` — текст обзора.

## Загрузка данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# установим параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], 
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])

# сохраним коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

**Функция для поочередного вывода таблиц `books`, `authors`, `publishers`, `ratings`, `reviews`**

In [3]:
table = [
    'books', 
    'authors', 
    'publishers', 
    'ratings', 
    'reviews'
]

for name in table:
    query = f'SELECT * FROM {name}'
    display(pd.io.sql.read_sql(query, con = engine))

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

## Количество книг, вышедших после 1 января 2000 год

In [4]:
query = '''
    SELECT COUNT(book_id)
    FROM books
    WHERE publication_date > '01.01.2000'
    '''
pd.io.sql.read_sql(query, con = engine)

count
0    819

**Вывод:** после 1 января 2000 года было издано 819 книг

## Количество обзоров и средняя оценка для каждой книги

In [5]:
query = '''
    SELECT b.book_id AS book_id,
           b.title AS title_book,
           COUNT(DISTINCT rev.review_id) AS reviews,
           AVG(rat.rating) AS avg_rating
    FROM books b
    LEFT JOIN reviews rev ON rev.book_id=b.book_id
    LEFT JOIN ratings rat ON rat.book_id=b.book_id
    GROUP BY b.book_id
    ORDER BY reviews desc
    LIMIT 5
    '''
pd.io.sql.read_sql(query, con = engine)        

book_id                                         title_book  reviews  \
0      948                            Twilight (Twilight  #1)        7   
1      497                          Outlander (Outlander  #1)        6   
2      207                                    Eat  Pray  Love        6   
3      299  Harry Potter and the Chamber of Secrets (Harry...        6   
4      302  Harry Potter and the Prisoner of Azkaban (Harr...        6   

   avg_rating  
0    3.662500  
1    4.125000  
2    3.395833  
3    4.287500  
4    4.414634

**Вывод:** наибольшее количество обзоров у книги `Twilight (Twilight #1)` со средним рейтингом 3.66

## Издательство, выпустившее наибольшее число книг толще 50 страниц. Исключение из анализа брошюр

In [6]:
query = '''
    SELECT pub.publisher,
           COUNT(b.book_id) AS books
    FROM books b 
    LEFT JOIN publishers pub ON pub.publisher_id=b.publisher_id
    WHERE b.num_pages > 50
    GROUP BY publisher
    ORDER BY books DESC
    LIMIT 1
    '''
pd.io.sql.read_sql(query, con = engine)            

publisher  books
0  Penguin Books     42

**Вывод:** 42 - ровно столько выпустило издательство `Penguin Books`. Оно же является абсолютным лидером по количеству выпущенных книг

## Автор с наивысшей средней оценкой книг, учитывая только книги с 50 и более оценками

In [7]:
query = '''
    SELECT author,
           AVG(sub.book_avg_rating) AS author_avg_rating
    FROM books b
    INNER JOIN (SELECT book_id,
                        AVG(rating) AS book_avg_rating
                FROM ratings 
                GROUP BY book_id
                HAVING COUNT(rating_id) >= 50) AS sub
    ON sub.book_id=b.book_id
    INNER JOIN authors aut ON aut.author_id = b.author_id
    GROUP BY author
    ORDER BY author_avg_rating DESC
    LIMIT 1
    '''
pd.io.sql.read_sql(query, con = engine)

author  author_avg_rating
0  J.K. Rowling/Mary GrandPré           4.283844

**Вывод:** наивысшая средняя оценка среди книг с 50 и более оценками у Джоан Роулинг и её коллеги Мэри Грандпре - 4.28

## Среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [8]:
query = '''
    SELECT ROUND(AVG(review_count),0)
    FROM (SELECT COUNT(review_id) as review_count
          FROM reviews
          WHERE username IN
                           (SELECT username
                           FROM ratings
                           GROUP BY username
                           HAVING COUNT(rating) > 50) 
    GROUP BY username) rev                          
    '''
pd.io.sql.read_sql(query, con = engine)

round
0   24.0

**Вывод:** как видим, для пользователей, оставивших 50 оценок и более, показатель среднего количества обзоров равен 24

## Общий вывод

На основе проведенной работы подытожим:

* после 1 января 2000 года было издано 819 книг;
* наибольшее количество обзоров у книги `Twilight (Twilight #1)` со средним рейтингом 3.66;
* 42 книги - ровно столько выпустило издательство `Penguin Books`. Оно же является абсолютным лидером по количеству выпущенных книг;
* наивысшая средняя оценка среди книг с 50 и более оценками у Джоан Роулинг и её коллеги Мэри Грандпре - 4.28;
* как видим, для пользователей, оставивших 50 оценок и более, показатель среднего количества обзоров равен 24.